File: iOS/Kalorie/Core/UseCases/RefreshFavouriteFoodUseCase.swift
//
//  RefreshFavouriteFoodUseCase.swift
//  Kalorie
//
//  Created by Josef Antoni on 19.09.2026.
//

import Foundation

protocol RefreshFavouriteFoodUseCaseProtocol {
    func callAsFunction(_ item: FoodItemDomain) async throws -> FoodItemDomain
}

struct RefreshFavouriteFoodUseCase: RefreshFavouriteFoodUseCaseProtocol {

    // MARK: - Properties

    private let dataProvider: any FirestoreDataProviderProtocol
    private let authProvider: any AuthProviderProtocol

    // MARK: - Init

    init(dataProvider: any FirestoreDataProviderProtocol, authProvider: any AuthProviderProtocol) {
        self.dataProvider = dataProvider
        self.authProvider = authProvider
    }

    // MARK: - Functions

    func callAsFunction(_ item: FoodItemDomain) async throws -> FoodItemDomain {
        guard item.kind == .catalogue else { return item }
        guard let userId = authProvider.userId else { throw AuthError.notAuthenticated }
        let catalogueDTO: FoodItemDTO? = try await dataProvider.loadAsync(id: item.id, from: Constants.Firestore.foodItems)
        guard let fresh = catalogueDTO?.asDomain(), fresh != item else { return item }
        let collection = Constants.Firestore.favouriteFoods(userId: userId)
        let existing: FavouriteFoodDTO? = try await dataProvider.loadAsync(id: item.id, from: collection)
        guard let existing else { return fresh }
        let dto = FavouriteFoodDTO(item: fresh, favouritedAt: Date(timeIntervalSince1970: existing.favouritedAt))
        try await dataProvider.setAsync(dto, id: item.id, in: collection)
        return fresh
    }
}

#if DEBUG
struct RefreshFavouriteFoodUseCaseFake: RefreshFavouriteFoodUseCaseProtocol {

    // MARK: - Properties

    var stubbedItem: FoodItemDomain?
    var shouldThrow = false

    // MARK: - Functions

    func callAsFunction(_ item: FoodItemDomain) async throws -> FoodItemDomain {
        if shouldThrow { throw URLError(.unknown) }
        return stubbedItem ?? item
    }
}
#endif

File: Android/app/src/main/java/antoni/kalorie/core/usecases/RefreshFavouriteFoodUseCase.kt
package antoni.kalorie.core.usecases

import antoni.kalorie.core.auth.AuthError
import antoni.kalorie.core.auth.AuthProviderProtocol
import antoni.kalorie.core.models.FoodItemDomain
import antoni.kalorie.core.models.FoodItemKind
import antoni.kalorie.core.networking.FavouriteFoodDTO
import antoni.kalorie.core.networking.FirestoreDataProviderProtocol
import antoni.kalorie.core.networking.FoodItemDTO
import antoni.kalorie.core.networking.loadAsync
import antoni.kalorie.core.networking.setAsync
import antoni.kalorie.core.utils.Constants
import antoni.kalorie.core.utils.instantFromEpochSeconds

interface RefreshFavouriteFoodUseCaseProtocol {
    suspend operator fun invoke(item: FoodItemDomain): FoodItemDomain
}

class RefreshFavouriteFoodUseCase(
    private val dataProvider: FirestoreDataProviderProtocol,
    private val authProvider: AuthProviderProtocol,
) : RefreshFavouriteFoodUseCaseProtocol {

    // MARK: - Functions

    override suspend fun invoke(item: FoodItemDomain): FoodItemDomain {
        if (item.kind != FoodItemKind.CATALOGUE) return item
        val userId = authProvider.userId ?: throw AuthError.NotAuthenticated
        val catalogueDTO: FoodItemDTO? = dataProvider.loadAsync(id = item.id, from = Constants.Firestore.FOOD_ITEMS)
        val fresh = catalogueDTO?.asDomain()
        if (fresh == null || fresh == item) return item
        val collection = Constants.Firestore.favouriteFoods(userId)
        val existing: FavouriteFoodDTO = dataProvider.loadAsync(id = item.id, from = collection) ?: return fresh
        val dto = FavouriteFoodDTO(item = fresh, favouritedAt = instantFromEpochSeconds(existing.favouritedAt))
        dataProvider.setAsync(dto, id = item.id, inCollection = collection)
        return fresh
    }
}
